feat(trainer): Support namespaced TrainingRuntime in the SDK

shaikmoeed opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines: https://www.kubeflow.org/docs/about/contributing
2. To know more about Kubeflow SDK, check the developer guide:
    https://github.com/kubeflow/sdk/blob/main/CONTRIBUTING.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
Add support to list/get namespaced TrainingRuntime.

**Which issue(s) this PR fixes**:

Fixes #88 

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/) included if any changes are user facing

kramaranya commented:
/ok-to-test

abhijeet-dhumal commented on kubeflow/trainer/backends/kubernetes/backend.py:
```diff
@@ -88,33 +100,55 @@ def list_runtimes(self) -> list[types.Runtime]:
 
     def get_runtime(self, name: str) -> types.Runtime:
-        """Get the the Runtime object"""
+        """Get the the Runtime object prefer namespaced, fall-back to cluster-scoped"""
```
Same change goes for each occurence here

szaher commented on kubeflow/trainer/backends/kubernetes/backend_test.py:
```diff
@@ -490,6 +500,37 @@ def create_cluster_training_runtime(
+        ),
+    )
+
+
```
did you mean to create this in `kubernetes/backend_test.py`? 
this is not a test function and I believe it should be added to the `TrainerClient` and propagated to the different backends.

shaikmoeed replied:
Similar to [create_train_job](https://github.com/kubeflow/sdk/blob/main/kubeflow/trainer/backends/kubernetes/backend_test.py#L446), I thought to use `create_training_runtime` to pass [here ](https://github.com/shaikmoeed/sdk/blob/fix/namespace-trainingruntime-list/kubeflow/trainer/backends/kubernetes/backend_test.py#L361-L369) instead of empty list. Please correct me, if this was not intended?

kramaranya commented on kubeflow/trainer/backends/kubernetes/backend.py:
```diff
@@ -87,28 +99,56 @@ def list_runtimes(self) -> list[types.Runtime]:
 
         return result
 
     def get_runtime(self, name: str) -> types.Runtime:
```
Similar to list_runtimes(), either case will cause a full failure

abhijeet-dhumal commented on kubeflow/trainer/backends/kubernetes/backend.py:
```diff
@@ -59,21 +59,33 @@ def __init__(self, cfg: KubernetesBackendConfig):
                 return result
 
-            for runtime in runtime_list.items:
+            for runtime in namespace_runtime_list.items + cluster_runtime_list.items:
```
WDYT? @kramaranya @szaher ?

shaikmoeed commented:
> Hey, @shaikmoeed! Are you still working on this PR? Did you have a chance to address the feedback?

@kramaranya Haven't got a chance to address the feedback. Will update the PR by next week.

shaikmoeed commented:
> Hey @shaikmoeed! Are you still working on this PR? It would be great if we could include this feature in 0.3 release (we aim to cut the release this week)

@kramaranya  Sorry for the delay, I have addressed all the comments! Can you please provide your review?

abhijeet-dhumal commented:
Following up on @astefanutti's [comment](https://github.com/kubeflow/sdk/pull/130#discussion_r2584044912) about symmetry between list_runtimes() and get_runtime().

According to current behavior and if we consider separate scope arg : 
If Namespace has "torch-runtime", cluster also has "torch-runtime"
 client.list_runtimes() will return : 
```
- torch-runtime (scope=project)
- torch-runtime (scope=cluster)  ...duplicate name
```
client.get_runtime("torch-runtime")
Returns one:
```
- torch-runtime (scope=project)  .. namespaced wins
```
so user sees duplicate names in list, but get_runtime() silently picks one. 
This creates confusion — "which one will user actually get?"

so what if : 

client.list_runtimes()
Returns:
```
- torch-runtime (scope=project)  ... namespaced, shadows cluster
```
client.get_runtime("torch-runtime")
Returns:
```
- torch-runtime (scope=project)  ... matches what list shows
```

here are some edge cases i could think of on top of my head : 
- Same name in both scopes  -  Namespaced returned, cluster hidden 
- Only cluster runtime exists - Returned with scope="cluster"
- Only namespaced runtime exists - Returned with scope="project" 
- User needs shadowed cluster runtime - Use kubectl or rename .. rare admin case

WDYT? @shaikmoeed

shaikmoeed commented on kubeflow/trainer/types/types.py:
```diff
@@ -245,6 +245,13 @@ class TrainerType(Enum):
+    """Scope for training runtimes."""
+
+    NAMESPACE = "project"
+    CLUSTER = "cluster"
```
This is to stay inline with domain usage.
cc: @kramaranya
